pkg/repository/chartmuseum.go

```diff
@@ -0,0 +1,375 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kubebb/core/api/v1alpha1"
+)
+
+const minIntervalSeconds = 120
+
+var _ IWatcher = (*chartmuseum)(nil)
+
+func NewChartmuseum(
+	ctx context.Context,
+	logger logr.Logger,
+	c client.Client,
+	instance *v1alpha1.Repository,
+	statusLen int,
+	cancel context.CancelFunc) IWatcher {
+
+	duration := time.Duration(minIntervalSeconds) * time.Second
+	if instance.Spec.PullStategy != nil && instance.Spec.PullStategy.IntervalSeconds > minIntervalSeconds {
+		duration = time.Duration(instance.Spec.PullStategy.IntervalSeconds) * time.Second
+	}
+
+	if r := time.Duration(time.Second * minIntervalSeconds); r.Milliseconds() > duration.Milliseconds() {
+		logger.Info("the minimum cycle period is 120 seconds, but it is actually less, so the default 300 seconds is used as the period.")
+		duration = r
+	}
+	return &chartmuseum{
+		instance:  instance,
+		c:         c,
+		ctx:       ctx,
+		logger:    logger,
+		duration:  duration,
+		cancel:    cancel,
+		statusLen: statusLen,
+	}
+
+}
+
+type chartmuseum struct {
+	ctx       context.Context
+	cancel    context.CancelFunc
+	instance  *v1alpha1.Repository
+	duration  time.Duration
+	statusLen int
+
+	c      client.Client
+	logger logr.Logger
+}
+
+func (c *chartmuseum) Start() {
+	c.logger.Info("Start to fetch")
+	go wait.Until(c.Poll, c.duration, c.ctx.Done())
+}
+
+func (c *chartmuseum) Stop() {
+	c.logger.Info("Delete Or Update Repository, stop watcher")
+	c.cancel()
+}
+
+func (c *chartmuseum) Poll() {
+	cond := v1alpha1.Condition{
+		Status:             v1.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+	}
+	indexFile, err := c.fetchIndexYaml()
+	if err != nil {
+		c.logger.Error(err, "Failed to fetch index file")
+		cond.Status = v1.ConditionFalse
+		cond.Message = "failed to fetch index file"
+		cond.Reason = v1alpha1.ReasonUnavailable
+		cond.Type = v1alpha1.TypeFailedSync
+	} else {
+		cond.Message = "index yaml synced successfully, createing components"
+		cond.Reason = v1alpha1.ReasonAvailable
+		cond.Type = v1alpha1.TypeSynced
+
+		diffAction, err := c.diff(indexFile)
+		if err != nil {
+			c.logger.Error(err, "failed to get diff")
+			cond.Status = v1.ConditionFalse
+			cond.Message = err.Error()
+			cond.Reason = v1alpha1.ReasonUnavailable
+			cond.Type = v1alpha1.TypeFailedSync
+		} else {
+			for _, item := range diffAction[0] {
+				c.logger.Info("create component", "Component.Name", item.GetName(), "Component.Namespace", item.GetNamespace())
+				if c.Create(&item); err != nil && !errors.IsAlreadyExists(err) {
+					c.logger.Error(err, "failed to create")
+				}
+			}
+			for _, item := range diffAction[1] {
+				c.logger.Info("update component", "Component.Name", item.GetName(), "Component.Namespace", item.GetNamespace())
+				if err := c.Update(&item); err != nil {
+					c.logger.Error(err, "failed to update component status")
+				}
+			}
```

Comment:
for `deleted` components, change its condition to `Deleted` or `Deprecated` or else...  @Abirdcfly  any comments? seems this needs to be added at `Condition`.

Reply:
Add a filed? Currently component does not have a `Condition` field.

Reply:
👌  add a new field in `status` @Abirdcfly  any comments?